### PR TITLE
mgmt: hawkbit: Fix case where body_data not set correctly in respons_cb

### LINKED ISSUE #34760: Hawkbit not downloading large files
**Describe the bug**
When dowloading a large fil/update using the hawkbit client the file gets corrupted. This is caused by the hawkbit application not checking the start of the buffer correctly. After a quick gloss over the other code in 'response_cb' this also the case for other actions in the application.

**To Reproduce**
1. Build an application that uses hawkbit for FOTA updates
2. Launch the application on your board.
3. Download the binary (2 times larger than 'RECV_BUFFER_SIZE') from hawkbit
4. Compare the binary that is uploaded to the downloaded binary

Now the binary that is downloaded will differ from the one that is uploaded.

**Expected behavior**
The downloaded binaries match.

**Impact**
Showstopper. This problem is preventing me from continuing the project, since this is an essential part of the project.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK
 - Zephyr 2.4.0 (Also affects zephyr 2.5.0 and master branch)

**Additional context**
There are two locations to fix this. The first is in hawkbit.c:  `response_cb`

```
case HAWKBIT_DOWNLOAD:
		if (hb_context.dl.http_content_size == 0) {
			body_data = rsp->body_start;                       // <--- Not the only condition body_start is set
			body_len = rsp->data_len;
			body_len -= (rsp->body_start - rsp->recv_buf);
			hb_context.dl.http_content_size = rsp->content_length;
		}

		if ((rsp->body_found == 1) && (body_data == NULL)) {
			body_data = rsp->recv_buf;
			body_len = rsp->data_len;
		}
```

The error here is that body_start is not only set when the callback is first called, but can also be set when the receive buffer was not completely filled. This leads to reading `body_len` bytes from `rsp->recv_buf` instead off `rsp->body_start` wich always falls within
`rsp->recv_buf`. Fixing the bug in this location would require adding another check to see if `rsp->body_start` is set and setting
`body_start` accordingly. This would be my preferred solution.

The other solution would be in `http_client.c:http_wait_data`.
The fix here would probably be to always start receiving from the start of the buffer and only setting `rsp->body_start` when the start off the http body has been found. This would probably also maximize buffer utilization. The downside is this could break assumptions that have been made in applications that depend on this behavior. I don't know if this is the correct for the http client.

### PR DESCRIPTION
**Pull request by burumaj**
When downloading large files from the hawkbit server the only time
rsp->body_start is checked is on the first call. After this it is never checked again.
This caused the wrong data to be written to flash and a failed update.
This fixes  #34760

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

### CODE REVIEW

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 1]
```diff
@@ -795,8 +795,14 @@ static void response_cb(struct http_response *rsp,
 		if (hb_context.dl.http_content_size == 0) {
 			body_data = rsp->body_start;
 			body_len = rsp->data_len;
+			// subtract the size of the HTTP header from body_len
 			body_len -= (rsp->body_start - rsp->recv_buf);
 			hb_context.dl.http_content_size = rsp->content_length;
+		} else {
+			// more general case where body data is set, but no need to take the
+			// HTTP header into account
+			body_data = rsp->body_start;
```
Comment (Navin-Sankar) [inline subsys/mgmt/hawkbit/hawkbit.c:804]: if the receive buffer (body_start) and data_len is handled here means, 
then the next if statement is not required,
````            
 if ((rsp->body_found == 1) && (body_data == NULL)) {
            body_data = rsp->recv_buf;
            body_len = rsp->data_len;
}
````

Comment (burumaj) [reply subsys/mgmt/hawkbit/hawkbit.c:804]: The if statement is still required. There is still a chance that `rsp->body_start` is set to `NULL`. This is the case when the data to be read starts at rsp->recv_buffer. The code that receives data from the from the socket ([`http_wait_data()`](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/http/http_client.c#L411) doesn't necessarily place the received data at the start of the buffer.

**Review by Navin-Sankar — APPROVED** [review]
LGTM.

**Comment by Navin-Sankar** [comment]
@burumaj fix the build failure. https://github.com/zephyrproject-rtos/zephyr/pull/34766/checks?check_run_id=2506469290

**Comment by sylvioalves** [comment]
@burumaj, can you fix the code compliance check error?

**Review by shubhamkulkarni97 — APPROVED** [review]

**Comment by sylvioalves** [comment]
@Navin-Sankar I created this as we have no reply from original user contributor:
https://github.com/zephyrproject-rtos/zephyr/pull/37242

**Pull request closed without merging**